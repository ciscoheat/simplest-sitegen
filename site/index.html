<!-- build:title -->FullOO Home Page<!-- /build:title -->
<!-- build:content -->
<h1 class="text-center"><strong>DCI - Data Context Interaction</strong></h1>
<h2 class="text-center">A New Role-Based Paradigm for specifying collaborating objects</h2>
<hr />
<p>The essence of object orientation is that networks of collaborating objects work
  together to achieve a common goal. The common sense of object oriented
  programming should reflect this essence with code that specifies how the objects collaborate.
  Our industry has, unfortunately, chosen differently and code is commonly written in terms of
  classes. A class tells us everything about the properties of the individual objects that are its
  instances. It does not tell us anything about how these instances work together to achieve the system
  behavior.</p>
<p>The result is that our code does not reveal everything about how a system will work. This is
  clearly not satisfactory, and we need a new paradigm as the foundation for more expressive
  code. The DCI (Data-Context-Interaction) is such a paradigm. DCI separates a
  program into different perspectives where each perspective focuses on certain system
  properties. Code in the <em>Data</em> perspective specifies how information is represented by stand-alone objects.
  Code in the <em>Context</em> perspective specifies runtime networks of interconnected objects as  similar structures of interconnected <em>roles</em>. Code
  in the <em>Interaction</em> perspective specifies how the networked objects collaborate to achieve the
system behavior. </p>
<p>Trygve Reenskaug is the originator of DCI.  Jim Coplien is its leading contributor and prime mover in the advancing DCI from early theory to accepted practice. </p>
<p>Trygve's home pages are at <a href="https://folk.universitetetioslo.no/trygver/">https://folk.universitetetioslo.no/trygver/</a> with information about his new project: <a href="https://folk.universitetetioslo.no/trygver/themes/Personal/pp-index.html">https://folk.universitetetioslo.no/trygver/themes/Personal/pp-index.html</a>.</p>
<p>Jim's DCI pages are at <a href="http://www.gertrudandcope.com">http://www.gertrudandcope.com</a> and <a href="http://www.leansoftwarearchitecture.com">http://www.leansoftwarearchitecture.com</a>.</p>
<p>Also see the DCI mailing list <a href="https://groups.google.com/forum/#!forum/object-composition">object-composition</a>.</p>

<p>There is a <a href="/doku.php?id=FAQ">FAQ</a> for frequently asked questions.</p>
<!-- /build:content -->