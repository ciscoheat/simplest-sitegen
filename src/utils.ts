/**
 * Hash using djb2
 */
 export const hash = (value : string | Uint8Array) => {
	let hash = 5381;
	let i = value.length;

	if (typeof value === 'string') {
		while (i) hash = (hash * 33) ^ value.charCodeAt(--i)
	} else {
		while (i) hash = (hash * 33) ^ value[--i]
	}

	return (hash >>> 0).toString(36)
}
