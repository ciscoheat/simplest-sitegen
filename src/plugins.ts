import path from 'upath'
import fs from 'fs-extra'
import { parse, type HTMLElement } from 'node-html-parser'
import { compile } from 'sass'
import c from 'ansi-colors'
import matter from 'gray-matter'
import md from 'markdown-it'

import { log } from './utils.js'
import { type Context } from './index.js'
import { hash } from './utils.js'

const isAbsolute = (url : string) => /^(?:[a-z+]+:)?\/\//i.test(url)

const resolvePath = (url : string, basePath : string, srcFile : string) => {
  if(url.startsWith('/'))
    return path.join(basePath, url)
  else {
    return path.join(path.dirname(srcFile), url)
  }
}

const cssScripts = (root : ReturnType<typeof parse>) => srcScripts(root, 'link', 'href')
const jsScripts = (root : ReturnType<typeof parse>) => srcScripts(root, 'script', 'src')

const srcScripts = (root : ReturnType<typeof parse>, selector : string, attr : string) => {
  return root.querySelectorAll(selector)
    .filter((el : HTMLElement) => !!el.attributes[attr])
    .map((el : HTMLElement) => ({
      el,
      attr,
      file: el.attributes[attr]
    })) as {el: HTMLElement, attr: string, file: string}[]
}

const html = ['.html', '.htm']

/////////////////////////////////////////////////////////////////////

export const cacheBust = {
  parse: async (context : Context, srcFile : string, content : string) => {
    const root = parse(content, {comment: true})
    const scriptFiles = cssScripts(root).concat(jsScripts(root))
      .filter(f => !f.file.includes('?'))
      .filter(f => !isAbsolute(f.file))

    if(!scriptFiles.length) return content

    for (const {el, attr, file} of scriptFiles) {
      const inputPath = resolvePath(file, context.config.input, srcFile)
      try {
        const content = await fs.readFile(inputPath).catch(() => {
          // Test if output exists instead of input
          const outputPath = resolvePath(file, context.config.output, srcFile)
          return fs.readFile(outputPath)
        })
        el.setAttribute(attr, file + '?' + hash(content))
      } catch (e) {
        log(c.red('Warning: ') + file + ' not found, referenced in ' + srcFile)
      }
    }

    return root.toString()
  }
}

/////////////////////////////////////////////////////////////////////

let sass : typeof compile

export const compileSass = {
  parse: async (context : Context, srcFile : string, content : string) => {
    const root = parse(content, {comment: true})
    const cssRefs = cssScripts(root)

    if(!cssRefs.length) return content

    for (const link of cssRefs) {
      if(!(link.file.endsWith('.sass') || link.file.endsWith('.scss'))) continue
      if(!sass) sass = (await import('sass')).default.compile
      
      const compiled = sass(resolvePath(link.file, context.config.input, srcFile), context.config.sassOptions as any)
      const cssFileName = path.changeExt(link.file, '.css')

      await fs.outputFile(resolvePath(cssFileName, context.config.output, srcFile), compiled.css)
      link.el.setAttribute(link.attr, cssFileName)

      if(compiled.sourceMap?.file) {
        const mapPath = resolvePath(
          path.changeExt(link.file, '.css.map', undefined, 8), 
          context.config.output, 
          srcFile
        )
        await fs.outputFile(mapPath, compiled.sourceMap.file)
      }
    }

    return root.toString()
  }
}

/////////////////////////////////////////////////////////////////////

export const htmlFiles = {
  parse: async (context : Context, file : string, content : string) => {
    return content.includes('<!-- /build:content -->')
      ? context.parser.processContent(content)
      : content
  }
}

/////////////////////////////////////////////////////////////////////

export const markdown = {
  extensions: ['.md'],
  parse: async (context : Context, file : string, content : string) => {
    const frontMatter = matter(content)
    const parsed = md(context.config.markdownOptions as any).render(frontMatter.content)
    const vars = Object.entries(frontMatter.data).map(([key, value]) => `<!-- build:${key} -->${value}<!-- /build:${key} -->`)

    return {
      file: path.changeExt(file, '.html'), 
      content: vars.join("\n") + `\n<!-- build:content -->${parsed}<!-- /build:content -->`
    }
  }
}