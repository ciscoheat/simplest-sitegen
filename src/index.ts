import fs from 'fs-extra'
import debug from 'debug'
// @ts-ignore
import Templator from 'template-html'
import fg from 'fast-glob'
import path from 'upath'
import c from 'ansi-colors'
import { cwd } from 'process'
import { Stats } from 'fs'
import { Options } from 'browser-sync'

import { log } from './utils.js'
import { cacheBust, compileSass, htmlFiles, markdown } from './plugins.js'

interface Rename {
  file: string
  content : string | Uint8Array
}

interface Remove {
  file: string
  action: 'remove'
}

interface Keep {
  file: string
  action: 'keep'
}

type ParsedFile = string | Rename | Remove | Keep // | (Rename | Remove | Keep)[]

type FileParser = (context : Context, file : string, content : string) => Promise<ParsedFile>

type FilesPlugin = {
  extensions: string[],
  parse: FileParser
}

type HtmlParser = {
  template: string
  processContent(content : string) : string
}

const defaultConfig = {
  input: "src" as string,
  output: "build" as string,
  template: "template.html" as string,
  ignoreExtensions: [".sass", ".scss"] as string[],
  passThrough: [],
  devServerOptions: { ui: false, notify: false } as Options,
  sassOptions: {style: "compressed"},
  markdownOptions: {},
  plugins: [] as FilesPlugin[]
}

export type Config = typeof defaultConfig

export type Context = {
  config: Config,
  parser: HtmlParser
}

/////////////////////////////////////////////////////////////////////

const d = debug('simplest')

const start = async (config2? : Partial<Config>) => {
  const config = Object.assign({}, defaultConfig, config2)

  try {
    const userConfigFile = path.join(cwd(), 'simplest.config.js')
    await fs.access(userConfigFile)

    try {
      const userConfig = await import('file:///' + userConfigFile)
      Object.assign(config, userConfig.default)
    } catch (e) {
      log('Could not import simplest.config.js. Did you add "type": "module" to package.json?')
    }
  } catch(e) {
    // File doesn't exist, keep going
  }

  if(config2)
    Object.assign(config, config2)

  const baseTemplate = path.join(config.input, config.template)

  try {
    // Sanity checks
    await fs.access(baseTemplate)
  } catch(e) {
    throw new Error(`Template file ${baseTemplate} not found!`)
  }

  const parser = new Templator({
    preserveTree: true,
    templateFile: baseTemplate
  }) as HtmlParser

  const context = {
    config,
    parser
  }

  /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

  const isNewer = (src : string, dest : string) => Promise.all([fs.stat(src), fs.stat(dest)])
    .then(([src1, dest1]) => {
      const output : boolean = src1.mtimeMs > dest1.mtimeMs
      //d(`${path.basename(src)}: ${output ? 'newer than output' : 'NOT newer than output'}`)
      return output
    })
    .catch(e => { 
      //d(`${path.basename(src)}: output didn't exist`)
      return true
    })

  const hasExtension = (exts : string[]) => (input : string) => exts.some(ext => input.endsWith(`${ext}`))

  ///////////////////////////////////////////////////////////

  const parseAllFiles = (async (plugins : FilesPlugin[]) => {
    const NOT_PARSED = ''

    const fileList = await fg(path.join(config.input, `/**/*.*`))
    const passThroughFiles = new Set(await fg(config.passThrough.map(glob => path.join(config.input, glob))))
    const allFiles = new Map<string, string>(fileList.map(f => [f, NOT_PARSED]))
    const plugins2 = [...plugins]

    // Prevent template from being parsed, since it has already been parsed in its own plugins.
    // This also makes it cacheable, because otherwise it would be overwritten here.
    allFiles.delete(config.template)

    const outputFileName = (file : string) => path.join(config.output, file.slice(config.input.length))
    const writeFile = (file : string, content : string | Uint8Array) => fs.outputFile(outputFileName(file), content)

    const usePluginMap = new Map(plugins2.map(p => [p, hasExtension(p.extensions)]))
    const usePlugin = (plugin : FilesPlugin, file : string) => usePluginMap.get(plugin)!(file)
    
    const templateMap = new Map<string, string>()

    // Generate template map (dir => template)
    for (const file of allFiles.keys()) {
      const templatePath = path.dirname(file)
      const fileName = path.basename(file)

      if(fileName != config.template) continue

      const templateOutputFile = outputFileName(file)
      const templateChanged = await isNewer(file, templateOutputFile)

      let templateContent = await fs.readFile(file, {encoding: 'utf8'})

      for (const plugin of plugins) {
        if(!usePlugin(plugin, file)) continue

        const parsed = await plugin.parse(context, file, templateContent)
        if(typeof parsed !== 'string') continue

        templateContent = parsed
      }

      templateMap.set(templatePath, templateContent)
              
      let currentTemplate = null
      if(!templateChanged) {
        try {
          currentTemplate = await fs.readFile(templateOutputFile, {encoding: 'utf8'})
        } catch(e) {}
      }
  
      if(currentTemplate != templateContent) {
        fs.outputFile(templateOutputFile, templateContent)
      }

      // Remove the parsed template(s) from the list so it won't be parsed again
      allFiles.delete(file)
    }
        
    const parseFile = async (file : string) => {

      const templateFor = (file : string) => {
        let fileName = file
        while(file && file != '.') {
          file = path.dirname(file)
          if(templateMap.has(file)) return templateMap.get(file)!
        }
        throw new Error('Template not found for ' + fileName)
      }
      
      context.parser.template = templateFor(file)

      let content = allFiles.get(file) || await fs.readFile(file, {encoding: 'utf8'})
  
      for (const plugin of plugins2) {
        if(!usePlugin(plugin, file)) continue

        const parsed = await plugin.parse(context, file, content)

        if(typeof parsed === 'string') {
          content = parsed
        } else if(Array.isArray(parsed)) {
          // TODO: Multiple file actions
        } else if('content' in parsed) {
          allFiles.set(parsed.file, parsed.content.toString())
          allFiles.delete(file)
        } else if(parsed.action == 'remove') {
          allFiles.delete(parsed.file)
        } else if(parsed.action == 'keep') {
          // Pass through unmodified to the next plugin, or just copy at the end
        } else {
          throw new Error('Unknown parsed value for ' + file + ': ' + parsed)
        }  
      }

      if(allFiles.has(file))
        allFiles.set(file, content)
    }

    const ignoreFile = hasExtension(config.ignoreExtensions)

    for (const file of allFiles.keys()) {
      if(passThroughFiles.has(file) || ignoreFile(file)) continue
      await parseFile(file)
    }
    
    // Copy the parsed and remaining files
    for (const [file, content] of allFiles) {
      if(content) {
        writeFile(file, content)
      } else if(!ignoreFile(file)) {
        const output = outputFileName(file)
        
        if(await isNewer(file, output)) {
          await fs.ensureDir(path.dirname(output))
          await fs.copy(file, output)
        }
      }
    }
  })
  
  ///// Starting up /////////////////////////////////////////////////

  const run = () => parseAllFiles(config.plugins.concat([markdown, compileSass, cacheBust, htmlFiles]))

  return {context, run}
}

export const simplestBuild = async (config? : Partial<Config>) => {
  const build = await start(config)
  const config2 = build.context.config

  await fs.remove(config2.output)
  return build.run()
}

export const simplestWatch = async (config? : Partial<Config>) => {
  const build = await start(config)
  const config2 = build.context.config

  const runWatch = (file : string, root : string, stat : Stats) => {
    log('Updated: ' + file)
    build.run()
  }

  log('Watching for file changes in ' + c.blue(config2.input))

  const sane = (await import('sane')).default
  const watcher = sane(config2.input)

  watcher.on('change', runWatch)
  watcher.on('add', runWatch)
  watcher.on('delete', (file) => {
    log('Deleted: ' + file)
    fs.remove(path.join(config2.output, file))
  })

  await build.run()
  return watcher as any
}

export const simplestDev = async (config? : Partial<Config>) => {
  log('Starting dev server')
  const build = await start(config)
  const config2 = build.context.config

  await fs.ensureDir(config2.output)
  await simplestWatch(config)
      
  const options = Object.assign({
    server: config2.output,
    files: config2.output
  }, config2.devServerOptions)

  const bs = (await import('browser-sync')).default.create()
  return bs.init(options)
}
