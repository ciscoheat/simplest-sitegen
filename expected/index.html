<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>FullOO Home Page</title>
  <link rel="stylesheet" href="/CSS/chota.min.css?glxzhr">
  <link rel="stylesheet" href="/CSS/fulloo.css?1l3yxcu">
  <!-- build:styles -->
</head>
<body>
  <nav class="nav hide-pr" id="tabs">
    <div class="nav-center">
      <a href="/">Home</a>
      <a href="/Introduction/">Introduction</a>
      <a href="/FAQ/">FAQ</a>
      <a href="/Documents/">Documents</a>
      <a href="/Examples/">Examples</a>
      <a href="/Downloads/">Downloads</a>
      <a href="/Events/">Events</a>
      <a href="/oowiki.php">Wiki</a>
    </div>
  </nav>

  <main class="container">
    
<h1 class="text-center"><strong>DCI - Data Context Interaction</strong></h1>
<h2 class="text-center">A New Role-Based Paradigm for specifying collaborating objects</h2>
<hr />
<p>The essence of object orientation is that networks of collaborating objects work
  together to achieve a common goal. The common sense of object oriented
  programming should reflect this essence with code that specifies how the objects collaborate.
  Our industry has, unfortunately, chosen differently and code is commonly written in terms of
  classes. A class tells us everything about the properties of the individual objects that are its
  instances. It does not tell us anything about how these instances work together to achieve the system
  behavior.</p>
<p>The result is that our code does not reveal everything about how a system will work. This is
  clearly not satisfactory, and we need a new paradigm as the foundation for more expressive
  code. The DCI (Data-Context-Interaction) is such a paradigm. DCI separates a
  program into different perspectives where each perspective focuses on certain system
  properties. Code in the <em>Data</em> perspective specifies how information is represented by stand-alone objects.
  Code in the <em>Context</em> perspective specifies runtime networks of interconnected objects as  similar structures of interconnected <em>roles</em>. Code
  in the <em>Interaction</em> perspective specifies how the networked objects collaborate to achieve the
system behavior. </p>
<p>Trygve Reenskaug is the originator of DCI.  Jim Coplien is its leading contributor and prime mover in the advancing DCI from early theory to accepted practice. </p>
<p>Trygve's home pages are at <a href="https://folk.universitetetioslo.no/trygver/">https://folk.universitetetioslo.no/trygver/</a> with information about his new project: <a href="https://folk.universitetetioslo.no/trygver/themes/Personal/pp-index.html">https://folk.universitetetioslo.no/trygver/themes/Personal/pp-index.html</a>.</p>
<p>Jim's DCI pages are at <a href="http://www.gertrudandcope.com">http://www.gertrudandcope.com</a> and <a href="http://www.leansoftwarearchitecture.com">http://www.leansoftwarearchitecture.com</a>.</p>
<p>Also see the DCI mailing list <a href="https://groups.google.com/forum/#!forum/object-composition">object-composition</a>.</p>

<p>There is a <a href="/doku.php?id=FAQ">FAQ</a> for frequently asked questions.</p>

  </main>
  
  <script src="/js/highlight.min.js"></script>
  <script>
    // Navigation highlight
    const current = document.location.pathname
    document.querySelectorAll('#tabs a').forEach(a => {
      const tab = new URL(a.href).pathname

      if(tab == current) {
        a.classList.add('active')
      }
    })
  </script>    
</body>
</html>
